tests/core/fs/workspacefs/test_invalid_data.py

```diff
@@ -0,0 +1,112 @@
+# Parsec Cloud (https://parsec.cloud) Copyright (c) AGPLv3 2019 Scille SAS
+
+import pytest
+from pendulum import Pendulum
+
+from parsec.core.fs import FSError
+
+from tests.common import freeze_time
+
+
+@pytest.fixture
+async def testbed(running_backend, alice_user_fs, alice):
+    with freeze_time("2000-01-01"):
+        wid = await alice_user_fs.workspace_create("w1")
+        workspace = alice_user_fs.get_workspace(wid)
+        await workspace.sync()
+        local_manifest = await workspace.local_storage.get_manifest(wid)
+
+    class TestBed:
+        def __init__(self):
+            self._next_version = 2
+            self.defaults = {
+                "local_manifest": local_manifest,
+                "blob": None,
+                "signed_author": alice.device_id,
+                "backend_author": alice.device_id,
+                "signed_timestamp": Pendulum(2000, 1, 2),
+                "backend_timestamp": Pendulum(2000, 1, 2),
+                "author_signkey": alice.signing_key,
+                "key": workspace.get_workspace_entry().key,
+            }
+
+        async def run(self, exc_msg, **kwargs):
+            options = {**self.defaults, **kwargs}
+
+            if options["blob"] is None:
+                to_sync_um = options["local_manifest"].to_remote(
+                    author=options["signed_author"], timestamp=options["signed_timestamp"]
+                )
+                options["blob"] = to_sync_um.dump_sign_and_encrypt(
+                    author_signkey=options["author_signkey"], key=options["key"]
+                )
+
+            await running_backend.backend.vlob.update(
+                organization_id=alice.organization_id,
+                author=options["backend_author"],
+                encryption_revision=1,
+                vlob_id=wid,
+                version=self._next_version,
+                timestamp=options["backend_timestamp"],
+                blob=options["blob"],
+            )
+            self._next_version += 1
+
+            # This should trigger FSError
+            with pytest.raises(FSError) as exc:
+                await workspace.sync()
+            str(exc.value) == exc_msg
+
+            # Also test timestamped workspace
+            with pytest.raises(FSError) as exc:
+                await workspace.to_timestamped(options["backend_timestamp"])
+            str(exc.value) == exc_msg
```

Comment:
Maybe parametrize the `testbed` fixture and let the test catch the `FSError` exception? e.g:
```python
@pytest.mark.trio
async def test_empty_blob(testbed):
    with pytest.raises(FSError) as context:
        await testbed.run(blob=b"")
    assert context.value.message == "Cannot decrypt vlob: The nonce must be exactly 24 bytes long"
```

Reply:
That was exactly what I was doing before adding the timestamped part ^^

Before that we had only one call raising an exception so it was logical to let the caller test handle it
But with two errors being raised, I should have two methods (well 3 in fact: `init`, `run_workspace`, `run_timestamped_workspace`) that have to be called by the test with exception handling. This is much more cumbersome (especially considering the methods may not be easily idempotent) so I just KISS